File: lab2/gr9sub2/bosko_anton/src/mem_touch.py
#!/usr/bin/env python3
"""
mem_touch.py
Простая утилита для постепенного увеличения RSS (имитация роста потребления памяти).
Поддерживает:
 - --rss-mb TARGET    : цель (сколько MB удерживать)
 - --step-mb STEP     : шаг пошагового выделения (MB)
 - --sleep-ms N       : пауза между шагами (ms)
 - --limit-mb L       : (опционально) установить RLIMIT_AS в байтах до старта (ограничит адресное пространство)
Сигналы:
 - SIGUSR1 : добавить один STEP (увеличить цель на STEP)
 - SIGUSR2 : уменьшить цель на STEP (освободить последний шаг)
 - SIGTERM : graceful exit
Выводит свои VmSize/VmRSS из /proc/self/status каждый шаг.
"""

import argparse
import os
import signal
import sys
import time
import resource

chunks = []
stop_flag = False
target_mb = 0
step_mb = 64

def read_proc_status(pid):
    """Возвращает словарь ключ->значение для /proc/<pid>/status первых полезных полей."""
    status = {}
    try:
        with open(f"/proc/{pid}/status", "r") as f:
            for line in f:
                if ':' not in line:
                    continue
                k, v = line.split(':', 1)
                status[k.strip()] = v.strip()
    except FileNotFoundError:
        return {}
    return status

def print_mem(pid, note=""):
    st = read_proc_status(pid)
    vmrss = st.get("VmRSS", "n/a")
    vmsize = st.get("VmSize", "n/a")
    print(f"[mem_touch] pid={pid} target={target_mb}MB allocated_chunks={len(chunks)} step={step_mb}MB {note} | VmRSS={vmrss} VmSize={vmsize}", flush=True)

def sigterm_handler(signum, frame):
    global stop_flag
    print("[mem_touch] SIGTERM received, exiting gracefully", flush=True)
    stop_flag = True

def sigusr1_handler(signum, frame):
    global target_mb
    target_mb += step_mb
    print(f"[mem_touch] SIGUSR1: increasing target -> {target_mb}MB", flush=True)

def sigusr2_handler(signum, frame):
    global target_mb
    target_mb = max(0, target_mb - step_mb)
    # free one chunk immediately if present
    if chunks:
        freed = chunks.pop()
        del freed
    print(f"[mem_touch] SIGUSR2: decreasing target -> {target_mb}MB", flush=True)

def parse_args():
    p = argparse.ArgumentParser()
    p.add_argument("--rss-mb", type=int, default=128, dest="rss_mb", help="Initial target RSS (MB)")
    p.add_argument("--step-mb", type=int, default=32, dest="step_mb", help="Step to allocate per iteration (MB)")
    p.add_argument("--sleep-ms", type=int, default=200, dest="sleep_ms", help="Sleep between steps (ms)")
    p.add_argument("--limit-mb", type=int, default=0, dest="limit_mb", help="If >0, set RLIMIT_AS to this MB before allocation")
    return p.parse_args()

def set_rlimit_as(limit_mb):
    if limit_mb <= 0:
        return
    bytes_limit = limit_mb * 1024 * 1024
    try:
        resource.setrlimit(resource.RLIMIT_AS, (bytes_limit, bytes_limit))
        print(f"[mem_touch] RLIMIT_AS set to {limit_mb} MB ({bytes_limit} bytes)", flush=True)
    except Exception as e:
        print(f"[mem_touch] Failed to set RLIMIT_AS: {e}", flush=True)

def allocate_one_step(step_mb):
    # allocate but keep reference so memory is not freed
    try:
        size = step_mb * 1024 * 1024
        chunks.append(bytearray(size))
        return True
    except MemoryError:
        print("[mem_touch] MemoryError on allocation", flush=True)
        return False
    except Exception as e:
        print(f"[mem_touch] Allocation exception: {e}", flush=True)
        return False

def free_all():
    global chunks
    chunks = []
    # suggest GC
    import gc
    gc.collect()

def main():
    global target_mb, step_mb, stop_flag

    args = parse_args()
    target_mb = args.rss_mb
    step_mb = args.step_mb
    sleep_s = args.sleep_ms / 1000.0

    # optional RLIMIT_AS
    if args.limit_mb > 0:
        set_rlimit_as(args.limit_mb)

    signal.signal(signal.SIGTERM, sigterm_handler)
    signal.signal(signal.SIGINT, sigterm_handler)
    signal.signal(signal.SIGUSR1, sigusr1_handler)
    signal.signal(signal.SIGUSR2, sigusr2_handler)

    pid = os.getpid()
    print(f"[mem_touch] started pid={pid}, target={target_mb}MB step={step_mb}MB sleep={args.sleep_ms}ms", flush=True)
    print_mem(pid, note="initial")

    # initial allocation loop: allocate until we reach target
    while not stop_flag:
        allocated_mb = len(chunks) * step_mb
        if allocated_mb < target_mb:
            ok = allocate_one_step(step_mb)
            if not ok:
                # allocation failed — print diagnostics and sleep longer to avoid spinning
                print_mem(pid, note="after failed alloc")
                time.sleep(1.0)
            else:
                print_mem(pid, note="after alloc")
                time.sleep(sleep_s)
            continue
        elif allocated_mb > target_mb:
            # free one chunk
            if chunks:
                chunks.pop()
                print_mem(pid, note="after free one chunk")
            time.sleep(sleep_s)
            continue
        else:
            # allocated == target — idle until signal changes target
            print_mem(pid, note="at target (idle)")
            time.sleep(sleep_s)

    print("[mem_touch] exiting, freeing memory", flush=True)
    free_all()
    print("[mem_touch] done", flush=True)

if __name__ == "__main__":
    main()

File: lab2/gr9sub2/popova_darya/src/mem_touch.c
/* mem_touch.c
   Простой инструмент для постепенного увеличения RSS и наблюдения OOM/лимитов.
   Компиляция:
     gcc -O2 mem_touch.c -o mem_touch
*/
#define _GNU_SOURCE
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <unistd.h>
#include <signal.h>
#include <errno.h>
#include <sys/resource.h>
#include <sys/types.h>
#include <inttypes.h>

static volatile sig_atomic_t stop_flag = 0;
static volatile sig_atomic_t inc_flag = 0;
static volatile sig_atomic_t dec_flag = 0;

void on_sigterm(int s){ (void)s; stop_flag = 1; }
void on_sigusr1(int s){ (void)s; inc_flag = 1; }
void on_sigusr2(int s){ (void)s; dec_flag = 1; }

static void print_status(const char *tag) {
    // вывести VmSize и VmRSS из /proc/self/status (пара первых строк)
    FILE *f = fopen("/proc/self/status", "r");
    if (!f) { perror("fopen /proc/self/status"); return; }
    char line[256];
    printf("=== %s: /proc/self/status ===\n", tag);
    while (fgets(line, sizeof(line), f)) {
        if (strncmp(line, "VmSize:", 7) == 0 || strncmp(line, "VmRSS:", 6) == 0 || strncmp(line, "Uid:", 4) == 0) {
            fputs(line, stdout);
        }
    }
    fclose(f);
    fflush(stdout);
}

int main(int argc, char **argv){
    size_t target_mb = 512;
    size_t step_mb = 64;
    int sleep_ms = 200;
    size_t max_steps = 1000;
    size_t limit_as_mb = 0; // 0 — не ставим
    int opt;

    // простая парсинг аргументов
    for (int i=1;i<argc;i++){
        if (strcmp(argv[i],"--rss-mb")==0 && i+1<argc){ target_mb = (size_t)atoi(argv[++i]); }
        else if (strcmp(argv[i],"--step-mb")==0 && i+1<argc){ step_mb = (size_t)atoi(argv[++i]); }
        else if (strcmp(argv[i],"--sleep-ms")==0 && i+1<argc){ sleep_ms = atoi(argv[++i]); }
        else if (strcmp(argv[i],"--limit-as-mb")==0 && i+1<argc){ limit_as_mb = (size_t)atoi(argv[++i]); }
        else if (strcmp(argv[i],"--max-steps")==0 && i+1<argc){ max_steps = (size_t)atoi(argv[++i]); }
        else if (strcmp(argv[i],"--help")==0){ printf("Usage: %s [--rss-mb N] [--step-mb M] [--sleep-ms ms] [--limit-as-mb L]\n", argv[0]); return 0; }
    }

    // сигналы
    signal(SIGTERM, on_sigterm);
    signal(SIGINT, on_sigterm);
    signal(SIGUSR1, on_sigusr1);
    signal(SIGUSR2, on_sigusr2);

    // установить RLIMIT_AS если требуется
    if (limit_as_mb > 0) {
        struct rlimit rl;
        rl.rlim_cur = rl.rlim_max = (rlim_t)limit_as_mb * 1024 * 1024;
        if (setrlimit(RLIMIT_AS, &rl) != 0) {
            perror("setrlimit(RLIMIT_AS)");
            // продолжаем, но сообщим
        } else {
            printf("set RLIMIT_AS = %zu MB\n", limit_as_mb);
        }
    }

    // динамический список блоков
    size_t steps_allocated = 0;
    char **blocks = calloc(max_steps+4, sizeof(char*));
    if (!blocks) { perror("calloc"); return 2; }

    size_t target_steps = (target_mb + step_mb - 1) / step_mb;
    printf("Starting mem_touch: target_mb=%zu step_mb=%zu sleep_ms=%d (target_steps=%zu)\n",
           target_mb, step_mb, sleep_ms, target_steps);

    print_status("start");

    for (size_t iter = 0; !stop_flag && iter < 1000000; ++iter) {
        // handle user signals
        if (inc_flag) { target_steps += 1; inc_flag = 0; printf("SIGUSR1: increasing target_steps -> %zu\n", target_steps); }
        if (dec_flag && target_steps > 0) { target_steps -= 1; dec_flag = 0; printf("SIGUSR2: decreasing target_steps -> %zu\n", target_steps); }

        // allocate until we reach target
        while (steps_allocated < target_steps) {
            if (steps_allocated >= max_steps) { fprintf(stderr, "reached max_steps limit (%zu)\n", max_steps); break; }
            size_t bytes = step_mb * 1024UL * 1024UL;
            char *p = malloc(bytes);
            if (!p) {
                perror("malloc");
                print_status("malloc-failed");
                // при неудаче — завершаем с кодом 3 (показывает OOM/limit)
                fprintf(stderr, "Allocation failed at step %zu (requested %zu MB total)\n", steps_allocated, (steps_allocated*step_mb));
                free(blocks);
                return 3;
            }
            // touch memory чтобы гарантировать RSS увеличился
            for (size_t i = 0; i < bytes; i += 4096) p[i] = 0xA5;
            blocks[steps_allocated++] = p;
            printf("[mem_touch] allocated step %zu -> total_alloc_steps=%zu (%zu MB)\n",
                   steps_allocated, steps_allocated, steps_allocated * step_mb);
            print_status("after-alloc");
            // небольшой небольшй паузу — чтобы можно было снимать pidstat между шагами
            usleep(1000*50);
        }

        // если переполнено — ждем
        // sleep для наблюдения
        if (sleep_ms > 0) usleep(sleep_ms * 1000);

        // если нужно уменьшить цель — освобождаем последние блоки
        while (steps_allocated > target_steps) {
            char *p = blocks[--steps_allocated];
            if (p) { free(p); blocks[steps_allocated] = NULL; printf("[mem_touch] freed one step -> steps=%zu\n", steps_allocated); print_status("after-free"); }
        }

        // периодический вывод
        if (iter % 5 == 0) print_status("loop");

        // cooperative exit check
        if (stop_flag) break;
    }

    printf("mem_touch exiting; freeing %zu blocks\n", steps_allocated);
    for (size_t i = 0; i < steps_allocated; ++i) if (blocks[i]) free(blocks[i]);
    free(blocks);
    print_status("end");
    return 0;
}

